### PR TITLE
Let the UGH implement the UGH-API interface

### PR DESCRIPTION
**Pull request by matthias-ronge**
This will add the interface to the UGH.

After being merged, the SLUB Nexus needs to be updated.

### CODE REVIEW

File: src/main/java/ugh/LegacyUghImplementation.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
+/*
+ * (c) Kitodo. Key to digital objects e. V. <contact@kitodo.org>
+ *
+ * This file is part of the Kitodo project.
+ *
+ * It is licensed under GNU General Public License version 3 or later.
```
Comment (henning-gerhardt) [inline src/main/java/ugh/LegacyUghImplementation.java]: Most files in this project are licensed under the LGPL (Lesser GNU General Public License) so this file needs even be licensed under this license.

File: src/main/java/ugh/LegacyUghImplementation.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+ *
+ * For the full copyright and license information, please read the
+ * GPL3-License.txt file that was distributed with this source code.
```
Comment (henning-gerhardt) [inline src/main/java/ugh/LegacyUghImplementation.java]: You must adjust this as the mentioned file 'GPL3-License.txt' did not exists on this repo nor is this file correct for this project as the project is more or less licensed under LGPL.

File: src/main/java/ugh/fileformats/mets/MetsMods.java [hunk 3]
```diff
@@ -112,18 +108,48 @@
...
-import ugh.exceptions.MetadataTypeNotAllowedException;
 import ugh.exceptions.MissingModsMappingException;
-import ugh.exceptions.PreferencesException;
-import ugh.exceptions.ReadException;
-import ugh.exceptions.TypeNotAllowedAsChildException;
-import ugh.exceptions.TypeNotAllowedForParentException;
-import ugh.exceptions.UGHException;
-import ugh.exceptions.WriteException;
+
+/*******************************************************************************
```
Comment (henning-gerhardt) [inline src/main/java/ugh/fileformats/mets/MetsMods.java]: Why are you adding the license header again?

File: src/main/java/ugh/fileformats/mets/MetsModsImportExport.java [hunk 4]
```diff
@@ -71,12 +78,30 @@
...
 import ugh.dl.Person;
 import ugh.dl.Prefs;
-import ugh.exceptions.DocStructHasNoTypeException;
 import ugh.exceptions.ImportException;
-import ugh.exceptions.MetadataTypeNotAllowedException;
-import ugh.exceptions.PreferencesException;
-import ugh.exceptions.ReadException;
-import ugh.exceptions.WriteException;
+
+/*******************************************************************************
```
Comment (henning-gerhardt) [inline src/main/java/ugh/fileformats/mets/MetsModsImportExport.java]: Why are you adding the license header again?

File: src/main/java/ugh/script/UghConvert.java [hunk 5]
```diff
@@ -460,16 +454,10 @@ public static int testmain(String args[]) {
...
-            }
-        } else {
-            System.out.println("?WRITE  ERROR\nREADY.");
-            System.out.println(convertTo + " file '" + outputFile.getAbsolutePath() + "' could not be written");
-            return 16;
+        if (!commandLine.hasOption('q')) {
+            System.out.println(convertTo + " file '" + outputFile.getAbsolutePath() + "' written");
         }
-
+
```
Comment (henning-gerhardt) [inline src/main/java/ugh/script/UghConvert.java]: Adding white spaces on an empty line - why?

File: src/main/java/converter/Conversion/StarterMetaDataConversion.java [hunk 6]
```diff
@@ -91,10 +92,10 @@ public static void main(String[] args) throws Exception {
                 logger.debug("loading:'" + newFile.getAbsolutePath() + "' as RDF in ugh");
 
                 try {
-                    if (rdfInput.read(procFile.getAbsolutePath())) {
+                    rdfInput.read(procFile.getAbsolutePath());
```
Comment (henning-gerhardt) [inline src/main/java/converter/Conversion/StarterMetaDataConversion.java]: You are removing the if condition but previous condition block is now not correct indent (one level to deep)

File: src/main/java/converter/Conversion/StarterMetaDataConversion.java [hunk 7]
```diff
@@ -135,18 +129,11 @@ public static void main(String[] args) throws Exception {
...
                 if (!flagError) {
                     try {
-                        if (!metsOutput.write(procFile.getAbsolutePath())) {
-
-                            logger.error("File " + procFile.getAbsolutePath() + " couldn't be written in mets format");
-
-                            rollbackLog.info(newFile.getAbsolutePath() + " - Mets couldn't be saved - processing cancelled");
-                            flagError = true;
-                        } else {
+                        ((FileformatInterface) metsOutput).write(procFile.getAbsolutePath());
```
Comment (henning-gerhardt) [inline src/main/java/converter/Conversion/StarterMetaDataConversion.java]: Same here: removing if condition without adjusting the indent of previous condition block.

File: src/main/java/gov/loc/mets/Helper.java [hunk 8]
```diff
@@ -45,6 +35,15 @@
 import org.w3c.dom.Document;
 import org.w3c.dom.Node;
 
+import gov.loc.mets.DivType.Fptr;
```
Comment (henning-gerhardt) [inline src/main/java/gov/loc/mets/Helper.java]: Why are you moving the gov import statements to the end of import statements? Without your changes import statements was sorted by alphabet.

File: src/main/java/gov/loc/mets/HelperTest.java [hunk 9]
```diff
@@ -20,16 +18,17 @@
...
-import gov.loc.mets.MetsType.MetsHdr.Agent;
+package gov.loc.mets;
 
 import java.io.File;
 import java.io.IOException;
 import java.util.HashMap;
 
 import org.apache.xmlbeans.XmlOptions;
 
+import gov.loc.mets.MetsType.MetsHdr;
```
Comment (henning-gerhardt) [inline src/main/java/gov/loc/mets/HelperTest.java]: Same again: moving import statements from alphabetical ordering to mixed ordering.

File: src/main/java/ugh/dl/DocStruct.java [hunk 10]
```diff
@@ -2431,63 +2441,57 @@ public boolean addChild(Integer index, DocStruct inchild) throws TypeNotAllowedA
...
-            return false;
+            return;
         }
 
         // get next position of index
         int next = where.indexOf(44) + 1;
 
-        return next != 0 ? children.get(Integer.parseInt(where.substring(0, next - 1))).addChild(where.substring(next),
-                inchild) : addChild(Integer.valueOf(where), inchild);
+        if( next != 0 ){ ((DocStruct) children.get(Integer.parseInt(where.substring(0, next - 1)))).addChild(where.substring(next),
```
Comment (henning-gerhardt) [inline src/main/java/ugh/dl/DocStruct.java]: Please use a proper formatting for this if condition.

File: src/main/java/ugh/fileformats/mets/MetsMods.java [hunk 11]
```diff
@@ -112,18 +108,48 @@
...
+ * This Library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library. If not, see <http://www.gnu.org/licenses/>.
+ ******************************************************************************/
+
+import gov.loc.mets.AmdSecType;
```
Comment (henning-gerhardt) [inline src/main/java/ugh/fileformats/mets/MetsMods.java]: Why did you moved this gov import statements out of alphabetical ordering?

File: src/main/java/ugh/fileformats/mets/MetsModsImportExport.java [hunk 12]
```diff
@@ -71,12 +78,30 @@
...
+ * This Library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library. If not, see <http://www.gnu.org/licenses/>.
+ ******************************************************************************/
+
+import gov.loc.mods.v3.ModsDocument;
```
Comment (henning-gerhardt) [inline src/main/java/ugh/fileformats/mets/MetsModsImportExport.java]: Why did you move this import statement out of alphabetical ordering?

File: src/main/java/ugh/LegacyUghImplementation.java (hunk 1 above)
Comment (matthias-ronge) [reply src/main/java/ugh/LegacyUghImplementation.java]: I don’t like the idea to design my own license header. Would it be okay to use the GDZ header as it is in all other files?

File: src/main/java/ugh/LegacyUghImplementation.java (hunk 2 above)
Comment (matthias-ronge) [reply src/main/java/ugh/LegacyUghImplementation.java]: I don’t like the idea to design my own license header. Would it be okay to use the GDZ header as it is in all other files?

File: src/main/java/ugh/LegacyUghImplementation.java (hunk 1 above)
Comment (henning-gerhardt) [reply src/main/java/ugh/LegacyUghImplementation.java]: This should be ok. If we would change it, then we change all of them to the new header.

File: src/main/java/ugh/LegacyUghImplementation.java (hunk 2 above)
Comment (henning-gerhardt) [reply src/main/java/ugh/LegacyUghImplementation.java]: See my answer above.

File: src/main/java/ugh/fileformats/mets/MetsMods.java (hunk 3 above)
Comment (matthias-ronge) [reply src/main/java/ugh/fileformats/mets/MetsMods.java]: Eclipse treats the license header as comment of an import, when it is misplaced below the package declaration, which was the case in this file. this is the original license header which was “sorted” here by Eclipse. Since the imports are collapsed by default, I didn’t see that it was moved down here, neither did one of my reviewers. I got the review comment why I had “deleted” the license header, which which was not so true. So I added one on top of the file. The one on top is the new one, this one is the original one. I will fix this. This response is only for your understanding.

File: src/main/java/ugh/fileformats/mets/MetsModsImportExport.java (hunk 4 above)
Comment (matthias-ronge) [reply src/main/java/ugh/fileformats/mets/MetsModsImportExport.java]: Same story here.

File: src/main/java/ugh/script/UghConvert.java (hunk 5 above)
Comment (matthias-ronge) [reply src/main/java/ugh/script/UghConvert.java]: Eclipse magic … I don’t know. Was fixed.

File: src/main/java/converter/Conversion/StarterMetaDataConversion.java (hunk 6 above)
Comment (matthias-ronge) [reply src/main/java/converter/Conversion/StarterMetaDataConversion.java]: I found that the Eclipse function to corret indents seems not to work reliably. If I select the whole file (<kbd>Ctrl</kbd>+<kbd>A</kbd>) it works, if not, it does nothing. I don’t know if this is the desired behaviour, at least, it is confusing, as opposed to source code formatter. Was Fixed.

File: src/main/java/converter/Conversion/StarterMetaDataConversion.java (hunk 7 above)
Comment (matthias-ronge) [reply src/main/java/converter/Conversion/StarterMetaDataConversion.java]: Same as above.

File: src/main/java/gov/loc/mets/Helper.java (hunk 8 above)
Comment (matthias-ronge) [reply src/main/java/gov/loc/mets/Helper.java]: Having import statements in alphabetical order with empty lines between the blocks requires some manual configuration in Eclipse. All package groups must be configured manually, all other packages go below there. When I set this up, I did a grep for `^import ([^.]+)` on all the kitodo-production *.java files, and added the found package groups in alphabetical order. However, the package starting with `gov` was not part of kitodo-production so I did not add it. I have now updated the configuration.

File: src/main/java/ugh/fileformats/mets/MetsModsImportExport.java (hunk 12 above)
Comment (matthias-ronge) [reply src/main/java/ugh/fileformats/mets/MetsModsImportExport.java]: Same as above.

File: src/main/java/ugh/fileformats/mets/MetsMods.java (hunk 11 above)
Comment (matthias-ronge) [reply src/main/java/ugh/fileformats/mets/MetsMods.java]: Same as above.
